`ultralytics 8.1.46` add TensorRT 10 support

Burhan-Q opened the pull request:
Working to add TensorRT >=10 support now that it's in early access. TensorRT 10 [documentation reference](https://docs.nvidia.com/deeplearning/tensorrt/archives/#trt_10). Appending code to add support for latest version changes, but retains code for use with prior versions of TensorRT.

- [x] update `ultralytics/engine/exporter.py`
  - tested on Windows 10 with `TensorRT 10.0.0b6`, `Ultralytics 8.1.42`, `PyTorch 2.2.0`, `ONNX 1.15.0`
  - [x] `model.export(format="engine")`
  - [x] `model.export(format="engine", half=True, dynamic=True)`
  - [x] `model.export(format="engine", half=True, dynamic=True, simplify=True, workspace=5, imgsz=320)`
  - [x] verify compatibility with `TensorRT <10` still functional
- [x] update `ultralytics/AutoBackend`
  - [x] Standard export
  - [x] `half` + `dynamic` export
  - [x] `half` + `dynamic` + `simplify` + `imgsz=320` export
  - [x] verify still compatible with older TensorRT model exports

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://ultralytics.com) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. **Check for Existing Contributions**: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. **Link Related Issues**: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. **Elaborate Your Changes**: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. **Ultralytics Contributor License Agreement (CLA)**: To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    _I have read the CLA Document and I sign the CLA_

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved support for NVIDIA TensorRT exports in Ultralytics YOLOv8 models.

### 📊 Key Changes
- Added a new test for exporting YOLO models to the NVIDIA TensorRT format.
- Updated Ultralytics package version to 8.1.46.
- Implemented support for exporting models to TensorRT, including dynamic input sizes and FP16 precision.
- Added compatibility with the latest TensorRT version 10 changes, including modifications to workspace settings and serialized network building.
- Enhanced the error handling and logging for TensorRT model context creation to guide users on version compatibility.
- Adjusted TensorRT export and model loading code to support both TensorRT versions below and above 10, focusing on memory management and dynamic input handling.

### 🎯 Purpose & Impact
- **Enhanced Exporting**: This update enables more efficient use of the YOLO model by allowing it to be exported to TensorRT, thereby leveraging NVIDIA's optimizations for faster inference on compatible hardware. 🚀
- **Broader Compatibility**: The changes ensure the model exportation is compatible with a wider range of TensorRT versions, including the latest release. This means users with newer NVIDIA GPUs can harness improved performance without compatibility issues. 🔧
- **Dynamic and Precision Support**: By supporting dynamic input sizes and FP16 precision, models can handle varying image sizes more effectively and run faster on supported devices, making it more flexible and efficient for real-time applications. 🔄
- **Error Handling and Guidance**: Enhanced error messages help users diagnose and understand compatibility issues between the exported models and their TensorRT versions, making troubleshooting easier. 🛠️

These updates are aimed at improving the user experience by ensuring compatibility with newer technologies and optimizing model performance, marking a significant step forward in deploying Ultralytics YOLO models in production environments.

Burhan-Q committed:
add support for TensorRT 10

only modifies existing code to keep prior functionalities in place

Burhan-Q committed:
add TensorRT10 compatability

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q commented:
Not happy with how this is at present, but it's a _reasonable_ starting point that doesn't impact inference speeds. When testing with TRT 9 exports with TRT 10 environment, the TRT 9 model would not load and would return `None` when `model.create_execution_context()` is called. The same occurs when attempting to use TRT 10 exports with TRT 9 environment (not really surprising). 

[TensorRT python API docs](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/)

Need to do more investigation into TRT version compatibility.

glenn-jocher commented:
@Burhan-Q got it! 

You can also check version like this instead of using Try: Except fields:

```python
if check_version(trt.__version__, "<10.0.0"):
    # logic here
```

Or you could return the Major version like this: 

```python
trt_major_version = parse_version(trt.__version__)[0]
```

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Update exporter.py

Burhan-Q reviewed (commented):
@glenn-jocher I saw you added some updates, I was thinking about using the walrus operator to assign `is_trt_10`

```py
_ = (is_trt_10 := check_version(trt.__version__, ">=10.0.0"))
```

Thought this could be really clean. I'll let you make the call on how to design it tho. Let me know if you want me to test anything on Windows or Linux

glenn-jocher committed:
Updates

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

ZouJiu1 committed:
Add support to exporting or inference with TensorRT 10.0.0b6, and fix bugs when exporting a tensorrt file .engine with flag half=True or when inference (#9840)

Co-authored-by: UltralyticsAssistant <web@ultralytics.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented:
@Burhan-Q oh hey buddy, `:=` is nice but not supported until Python 3.9 I think, so for 3.8 compatibility for the rest of the year or so we can't introduce these. Next year though!

glenn-jocher commented:
@Burhan-Q ok I've added a bunch of updates from @ZouJiu1 including for AutoBackend forward call. Can you review to make sure I didn't break anything? If it looks all good I think we can merge this for our next release.

glenn-jocher committed:
Update __init__.py

Burhan-Q commented:
My bad, I thought I read it was supported in 3.8 as well

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher commented:
@Burhan-Q yeah actually maybe you're right, I got myself mixed up. So even though we're officially 3.7 deprecated we still see a high amount of workflows out there running 3.7 per our analytics so we have to try to stay 3.7-friendly for the time being.

glenn-jocher commented:
@Burhan-Q ok just merged 8.1.45 and queued this for 8.1.46. Let's aim to merge this PR Monday night if all goes well.

Burhan-Q commented:
Sounds good to me! Thanks @glenn-jocher

Burhan-Q reviewed (commented):
This fails with latest updates (tested okay previously) with TensorRT 10. The combination of `(half=True, dynamic=True)` is the failure mode so far, if only one argument is present export seems to work okay.  Will do testing tomorrow with TensorRT 9 exports and inference.

```py
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
out = model.export(format="engine", half=True, dynamic=True)
```

```
WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0

Ultralytics YOLOv8.1.46 🚀 Python-3.11.6 torch-2.2.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3080, 12288MiB)
YOLOv8n summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from 'yolov8n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (6.2 MB)

ONNX: starting export with onnx 1.15.0 opset 17...
ONNX: export failure ❌ 2.7s: "upsample_nearest2d_channels_last" not implemented for 'Half'
TensorRT: export failure ❌ 2.7s: "upsample_nearest2d_channels_last" not implemented for 'Half'

Traceback (most recent call last):

  File "<stdin>", line 1, in <module>
  File "ultralytics/ultralytics/engine/model.py", line 601, in export
    return Exporter(overrides=args, _callbacks=self.callbacks)(model=self.model)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/exporter.py", line 286, in __call__
    f[1], _ = self.export_engine()
              ^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/exporter.py", line 139, in outer_func
    raise e
  File "ultralytics/ultralytics/engine/exporter.py", line 134, in outer_func
    f, model = inner_func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/exporter.py", line 658, in export_engine
    f_onnx, _ = self.export_onnx()  # run before trt import https://github.com/ultralytics/ultralytics/issues/7016
                ^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/exporter.py", line 139, in outer_func
    raise e
  File "ultralytics/ultralytics/engine/exporter.py", line 134, in outer_func
    f, model = inner_func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/exporter.py", line 376, in export_onnx
    torch.onnx.export(
  File "torch/onnx/utils.py", line 516, in export
    _export(
  File "torch/onnx/utils.py", line 1613, in _export
    graph, params_dict, torch_out = _model_to_graph(
                                    ^^^^^^^^^^^^^^^^
  File "torch/onnx/utils.py", line 1135, in _model_to_graph
    graph, params, torch_out, module = _create_jit_graph(model, args)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/onnx/utils.py", line 1011, in _create_jit_graph
    graph, torch_out = _trace_and_get_graph_from_model(model, args)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/onnx/utils.py", line 915, in _trace_and_get_graph_from_model
    trace_graph, torch_out, inputs_states = torch.jit._get_trace_graph(
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/jit/_trace.py", line 1296, in _get_trace_graph
    outs = ONNXTracedModule(
           ^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/jit/_trace.py", line 138, in forward
    graph, out = torch._C._create_graph_by_tracing(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/jit/_trace.py", line 129, in wrapper
    outs.append(self.inner(*trace_inputs))
                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1501, in _slow_forward
    result = self.forward(*input, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/nn/tasks.py", line 89, in forward
    return self.predict(x, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/nn/tasks.py", line 107, in predict
    return self._predict_once(x, profile, visualize, embed)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/nn/tasks.py", line 128, in _predict_once
    x = m(x)  # run
        ^^^^
  File "torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/module.py", line 1501, in _slow_forward
    result = self.forward(*input, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/modules/upsampling.py", line 157, in forward
    return F.interpolate(input, self.size, self.scale_factor, self.mode, self.align_corners,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/nn/functional.py", line 4001, in interpolate
    return torch._C._nn.upsample_nearest2d(input, output_size, scale_factors)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: "upsample_nearest2d_channels_last" not implemented for 'Half'
```

glenn-jocher commented on ultralytics/engine/exporter.py:
```diff
@@ -237,7 +237,7 @@ def __call__(self, model=None):
         for _ in range(2):
             y = model(im)  # dry runs
-        if self.args.half and onnx and self.device.type != "cpu":
+        if self.args.half and (onnx or engine) and self.device.type != "cpu":
```
@Burhan-Q failure might be because of this. @ZouJiu1 was saying that without this the inputs and outputs remains FP32 (but the inner layers are FP16 I think).

Can you try with and without this line?

Burhan-Q replied:
@glenn-jocher yes, it appears that commenting this line and using `if self.args.half and onnx and self.device.type != "cpu":` instead works for both TensorRT >= 10.0 and TensorRT < 10. I may have misunderstood, but I think that the inputs (at least) for TensorRT are supposed to be FP32 and that the conversion happens internally, but I can't find where I read this so it might not be correct.

ZouJiu1 replied:
hi, @Burhan-Q I think using "(onnx or engine)" is necessary, the conversion with 10.0.0b6 is not internal. Because when use the converted engine tensorrt file,  the codes use input dtype to judge FP32 or FP16, not the arg.half flag, in line [autobackend.py#L244](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L244), [autobackend.py#L249](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L249) and [autobackend.py#L240](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L240). If I use `if self.args.half and onnx and self.device.type != "cpu":` to convert engine file with 10.0.0b6, the converted engine input shows FP32 not FP16 when I debug into [autobackend.py#L244](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L244).

So, I think `if self.args.half and (onnx or engine) and self.device.type != "cpu":` is necessary. the conversion is not internal.

Burhan-Q replied:
@ZouJiu1 this single change causes the [error documented here](https://github.com/ultralytics/ultralytics/pull/9516#pullrequestreview-1985271931). I get what you're saying, but it doesn't work when this line is used. We can always update to get the functionality you're looking for after we have solid compatibility with new and older versions of TensorRT

glenn-jocher committed:
Merge branch 'main' into trt10

Burhan-Q commented:
So I have only done a few tests so far and I'm seeing odd results that I don't recall having when I tested last week. I'm going to checkout an older commit to see if these issues were present, but I don't recall these being an issue.

1. When using `dynamic=True` with TensorRT < 10 inference is throwing an error, this does not occur with TensorRT >= 10

```
Traceback (most recent call last):

  File "<stdin>", line 1, in <module>
  File "ultralytics/ultralytics/engine/model.py", line 452, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
               ^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/engine/predictor.py", line 216, in stream_inference
    self.setup_model(model)
  File "ultralytics/ultralytics/engine/predictor.py", line 297, in setup_model
    self.model = AutoBackend(
                 ^^^^^^^^^^^^
  File "torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "ultralytics/ultralytics/nn/autobackend.py", line 286, in __init__
    batch_size = bindings["images"].shape[0]  # if dynamic, this is instead max batch size
                 ~~~~~~~~^^^^^^^^^^

KeyError: 'images'
```

2. Speeds are all over the place and on first call, postprocess time is significantly high

```py
from ultralytics import YOLO

# after exporting and new python session for TRT10
model = YOLO("yolov8n.engine", task="detect")

# Call 01
results = model.predict() # default ASSETS inference
>>> Speed: 5.5ms preprocess, 2.0ms inference, 556.5ms postprocess per image at shape (1, 3, 640, 640)

# Call 02
results = model.predict()
>>> Speed: 2.5ms preprocess, 6.3ms inference, 2.1ms postprocess per image at shape (1, 3, 640, 640)

# Call 03
results = model.predict()
>>> Speed: 3.0ms preprocess, 5.0ms inference, 5.0ms postprocess per image at shape (1, 3, 640, 640)

# Call 04
results = model.predict()
>>> Speed: 4.5ms preprocess, 44.5ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
```

```py
# New python session and export for TRT10
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
out = model.export(format="engine", half=True, dynamic=True, simplify=True)
model = YOLO(out, task="detect")

# Call 01
results = model.predict()
>>> Speed: 5.0ms preprocess, 12.5ms inference, 2.5ms postprocess per image at shape (1, 3, 640, 640)

# Call 02
results = model.predict()
>>> Speed: 37.0ms preprocess, 33.4ms inference, 3.5ms postprocess per image at shape (1, 3, 640, 640)

# Call 03
results = model.predict()
>>> Speed: 2.0ms preprocess, 2.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)

# Call 04
results = model.predict()
>>> Speed: 3.0ms preprocess, 10.5ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)

# Call 05
results = model.predict()
>>> Speed: 4.0ms preprocess, 102.8ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
```

Burhan-Q commented:
Yeah, @glenn-jocher reverting to [8418063](https://github.com/ultralytics/ultralytics/pull/9516/commits/8418063367fded5883d753ecf559eb0e2a89b515) doesn't throw any errors when running inference. The inference speeds are **much** better and more consistent, the first result shows quite slow speeds, but the rest are far better

```py
from ultralytics import YOLO
from ultralytics import ASSETS

im = ASSETS / "bus.jpg"
model = YOLO("yolov8n.engine", task="detect")

for _ in range(100):
    results = model.predict(im)

# TensorRT < 10
>>> Speed: 36.0ms preprocess, 1.0ms inference, 1430.3ms postprocess per image at shape (1, 3, 640, 640) # first call
>>> Speed: 3.0ms preprocess, 1.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 3.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 3.0ms preprocess, 2.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
...
>>> Speed: 2.0ms preprocess, 2.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640) # last call, most have nearly identical speeds

# TensorRT >= 10
>>> Speed: 27.0ms preprocess, 1.0ms inference, 1043.4ms postprocess per image at shape (1, 3, 640, 640) # first call
>>> Speed: 3.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 3.0ms preprocess, 2.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 2.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 3.0ms preprocess, 1.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)
>>> Speed: 2.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
...
>>> Speed: 2.0ms preprocess, 1.0ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640) # last call, most have nearly identical speeds

Burhan-Q commented:
I'm going to continue testing from https://github.com/ultralytics/ultralytics/pull/9516/commits/8418063367fded5883d753ecf559eb0e2a89b515

Burhan-Q committed:
Removed changes from 1f553c2 due to errors, now working and fixes issue with dynamic inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q committed:
ensure max shape dimension scales no smaller than 1

Burhan-Q commented:
Okay, I reverted to a start from [8418063](https://github.com/ultralytics/ultralytics/commit/8418063367fded5883d753ecf559eb0e2a89b515) as it removed _many_ issues. I tested (on Windows10) for all tasks using `tensorrt == 10.0.0b6` and `tensorrt == 9.2.0.post12.dev5`

| Task           | TensorRT >= 10 | TensorRT < 10 |
|:------------:|:---------------:|:--------------:|
| DETECT     |       ✔            |      ✔           |
| SEGMENT |       ✔            |      ✔           |
| POSE        |       ✔            |      ✔            | 
| OBB         |       ✔             |      ✔           |

All tasks exported without issue. I did discover an existing issue where `dynamic` didn't actually allow for any `imgsz` when exported. I fixed this by specifying `_min`, `_opt` (optimal), and `_max` inference dimensions. The `_min` value is hard set to `(1, 3, 32, 32)`, `_opt` (optimal) is based on the `imgsz` argument during export, and `_max` reuses the `workspace` argument to scale the max dimensions (by default 4x). This allows for the model to use _any_ image size for inference on the interval $[32, 4 * \text{imgsz})$ (default). 

After making this update, I found that the `dynamic` conditional wasn't triggering the logic for `dynamic` and would always throw an error about the wrong inference dimension size, so I changed the conditional from `and` to `or` so it would take this route when an input image was not the correct size. This uncovered a few more areas in `Autobackend` that were incompatible with TensorRT 10, and I added `try-except` logic to catch these.

Once done, I did not completely retest all tasks due to time constraints, but I checked a few examples to ensure both that there were no errors and that the new logic for `dynamic` sizing works. Additionally inference speeds are _significantly_ better. The first inference time does seem to be quite high, but running multiple iterations causes them to settle down with only a few spikes. Running repeated inference on a single image with a `for`-loop on a 100 iterations was seeing inference speeds around $1.0 \text{ms}$ and total end-to-end processing time of $2.0 - 3.0 \text{ms}$

I believe this is ready to merge unless there are any additional changes necessary or if there are any additional tests that should be run. If I get a chance later, I will also try to run this on Linux, but I don't expect there to be any compatibility issues. 🚀

Burhan-Q committed:
Merge branch 'trt10' of https://github.com/ultralytics/ultralytics into trt10

glenn-jocher committed:
Merge branch 'main' into trt10

Burhan-Q commented:
Testing branch with latest changes on Ubuntu 22.04 with RTX-2060

## TensorRT 10
- average 100 predict calls for `detect` $12.35\text{ ms}$
- average 100 predict calls for `segment` $6.44\text{ ms}$
- average 100 predict calls for `pose` $5.98\text{ ms}$
- average 100 predict calls for `obb` $4.86\text{ ms}$

## TensorRT 8.6.1.post1
- average 100 predict calls for `detect` $18.56 \text{ ms}$
- average 100 predict calls for `segment` $6.40  \text{ ms}$
- average 100 predict calls for `pose` $5.93 \text{ ms}$
- average 100 predict calls for `obb` $4.782 \text{ ms}$

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Updates

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Merge remote-tracking branch 'origin/trt10' into trt10

glenn-jocher commented:
@Burhan-Q this code block has too many if statements, making it hard to read for users. Can you refactor this into a single if statement as @ZouJiu1 had in his PR?

Also logic should be `is_trt10` to align with Exporter.

<img width="819" alt="Screenshot 2024-04-09 at 13 36 31" src="https://github.com/ultralytics/ultralytics/assets/26833433/57365c44-4eb5-4aee-9518-b09dd64f0dd0">

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
Refactor constants out of for loop for speed

glenn-jocher committed:
Update exporter.py

Burhan-Q committed:
refactored to use unified methods

Burhan-Q committed:
Merge branch 'trt10' of https://github.com/ultralytics/ultralytics into trt10

Burhan-Q commented:
@glenn-jocher removed excess conditionals from `Autobackend` and now using methods that are available in new and old versions of TensorRT

Burhan-Q committed:
remove redundant line

glenn-jocher committed:
remove self. as assigned automatically

glenn-jocher commented:
@Burhan-Q issue here


<img width="1329" alt="Screenshot 2024-04-09 at 17 15 51" src="https://github.com/ultralytics/ultralytics/assets/26833433/74b7d47f-4e38-4501-a900-de5fc87cae89">

glenn-jocher committed:
Add TensorRT export test

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Merge remote-tracking branch 'origin/trt10' into trt10

glenn-jocher commented:
@Burhan-Q I've added a TensorRT export and inference test in https://github.com/ultralytics/ultralytics/pull/9516/commits/3017bad20a41f0f2ef6ad8d195762fa05ee9b592

glenn-jocher committed:
Update test_cuda.py

Burhan-Q commented:
@glenn-jocher sounds good. Looks like the older TensorRT versions _require_ use of the `get_bindings` so I'm going to add these back in like you suggested using a singular `if-else` statement

Burhan-Q commented:
The oldest version of TensorRT I could install for python 3.10 is `tensorrt-8.5.3.1` does it make sense to continue support for anything older? I'm going to test out with python 3.8 to see if it will allow for an older version of TensorRT

```sh
pip install tensorrt==8.4.*

ERROR: Could not find a version that satisfies the requirement tensorrt==8.4.* 
(from versions: 0.0.1.dev5, 0.0.1, 8.5.1.7, 8.5.2.2, 8.5.3.1, 8.6.0, 8.6.1, 8.6.1.post1, 
9.0.0.post11.dev1, 9.0.0.post12.dev1, 9.0.1.post11.dev4, 9.0.1.post12.dev4, 
9.1.0.post11.dev4, 9.1.0.post12.dev4, 9.2.0.post11.dev5, 9.2.0.post12.dev5, 
9.3.0.post11.dev1, 9.3.0.post12.dev1, 10.0.0b6)
```

Burhan-Q commented:
@glenn-jocher it appears that the oldest version of TensorRT that is downloadable for install with python 3.8 is `8.5.1.7` and even with `8.5.3.1` for python 3.10, there seems to be an issue with `numpy` as it uses `np.bool` which was deprecated in `numpy==1.20` which might be pinned by another dependency.

Burhan-Q committed:
refactored to include older interface for `tensorrt<8.6`

Burhan-Q committed:
Merge branch 'trt10' of https://github.com/ultralytics/ultralytics into trt10

Burhan-Q committed:
refactor to include path for installs with `tensorrt<8.6`

Burhan-Q committed:
fix dynamic where self. is needed, align else, rearrange flow

Burhan-Q committed:
correct for error with TensorRT 8.4.3.1 without method `get_tensor_shape`

Burhan-Q commented:
@glenn-jocher CI is passing now and I was able to test with TensorRT `8.5.3.1` on python 3.8 (after forcing some old dependencies) using the latest commit. It looks like versions of TensorRT `8.4` can still be downloaded from the GitHub repo, but not installed from `pip`. According to the [release notes](https://docs.nvidia.com/deeplearning/tensorrt/release-notes/#tensorrt-8):

> APIs deprecated before TensorRT 8.4 will be removed in TensorRT 9.0.

so I guess this means that some components of the code are deprecated during each release. These are maintained for some time before being completely removed.

glenn-jocher committed:
delete redundant trt10 definition

glenn-jocher committed:
Align variable names in Exporter and Autobackend

glenn-jocher committed:
Merge branch 'main' into trt10

glenn-jocher committed:
move test to slow
